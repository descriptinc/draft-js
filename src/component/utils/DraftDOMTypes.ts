/**
 * (c) Facebook, Inc. and its affiliates. Confidential and proprietary.
 *
 * Types for things in the DOM used in Draft.js. These should eventaully be
 * added to the flow DOM lib itself.
 *
 * @emails oncall+draft_js
 */

// https://developer.mozilla.org/en-US/docs/Web/API/Selection
export type SelectionObject = {
  /**
   * Returns the Node in which the selection begins. Can return null if
   * selection never existed in the document (e.g., an iframe that was
   * never clicked on). */
  anchorNode: Node | null;
  anchorOffset: number;
  focusNode: Node | null;
  focusOffset: number;
  isCollapsed: boolean;
  rangeCount: number;
  type: string;

  removeAllRanges(): void;
  getRangeAt: (index: number) => Range;
  extend?: (node: Node, offset?: number) => void;
  addRange: (range: Range) => void;
  // ...etc. This is a non-exhaustive definition.
};
