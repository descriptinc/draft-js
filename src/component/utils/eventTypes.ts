export type SyntheticEvent<T = Record<string, unknown>> = React.SyntheticEvent<
  T
>;
export type SyntheticClipboardEvent<
  T = Record<string, unknown>
> = React.ClipboardEvent<T>;
export type SyntheticInputEvent = React.SyntheticEvent<HTMLInputElement>;
