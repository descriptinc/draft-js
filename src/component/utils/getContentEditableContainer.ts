/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+draft_js
 */

import invariant from '../../fbjs/invariant';
import isHTMLElement from './isHTMLElement';
import DraftEditor from '../base/DraftEditor.react';

export default function getContentEditableContainer(
  editor: DraftEditor,
): HTMLElement {
  const editorNode = editor.editorContainer;
  invariant(editorNode, 'Missing editorNode');
  invariant(
    isHTMLElement(editorNode.firstChild),
    'editorNode.firstChild is not an HTMLElement',
  );
  return editorNode.firstChild;
}
