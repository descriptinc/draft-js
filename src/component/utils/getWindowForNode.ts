/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+draft_js
 */

export default function getWindowForNode(node: Node | null): any {
  if (!node || !node.ownerDocument || !node.ownerDocument.defaultView) {
    return window;
  }
  return node.ownerDocument.defaultView;
}
