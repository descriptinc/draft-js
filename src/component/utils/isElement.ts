/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @emails oncall+draft_js
 */

export default function isElement(node: Node | undefined | null): boolean {
  if (!node || !node.ownerDocument) {
    return false;
  }
  return node.nodeType === Node.ELEMENT_NODE;
}
