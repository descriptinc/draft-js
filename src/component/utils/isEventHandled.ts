/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+draft_js
 */

import {DraftHandleValue} from '../../model/constants/DraftHandleValue';

/**
 * Utility method for determining whether or not the value returned
 * from a handler indicates that it was handled.
 */
export default function isEventHandled(value: DraftHandleValue): boolean {
  return value === 'handled' || value === true;
}
