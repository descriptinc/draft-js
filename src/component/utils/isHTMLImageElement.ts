/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @emails oncall+draft_js
 */
import isElement from './isElement';

export default function isHTMLImageElement(
  node?: Node,
): node is HTMLImageElement {
  if (!node || !node.ownerDocument) {
    return false;
  }
  return isElement(node) && node.nodeName === 'IMG';
}
