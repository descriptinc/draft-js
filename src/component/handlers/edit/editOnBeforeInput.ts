/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @flow strict-local
 * @emails oncall+draft_js
 */

'use strict';

import UserAgent from 'fbjs/lib/UserAgent';
import setImmediate from 'fbjs/lib/setImmediate';
import {
  EditorState,
  getBlockTree,
  getCurrentInlineStyle,
  pushContent,
  setEditorState,
} from '../../../model/immutable/EditorState';
import {DraftInlineStyle} from '../../../model/immutable/DraftInlineStyle';
import DraftModifier from '../../../model/modifier/DraftModifier';
import DraftEditor from '../../base/DraftEditor.react';
import isEventHandled from '../../utils/isEventHandled';
import {
  getStartOffset,
  isCollapsed,
} from '../../../model/immutable/SelectionState';
import getEntityKeyForSelection from '../../../model/entity/getEntityKeyForSelection';
import isSelectionAtLeafStart from '../../selection/isSelectionAtLeafStart';
import {some, zip} from '../../../model/descript/Iterables';
import {nullthrows} from '../../../fbjs/nullthrows';
import {SyntheticInputEvent} from '../../utils/eventTypes';

// When nothing is focused, Firefox regards two characters, `'` and `/`, as
// commands that should open and focus the "quickfind" search bar. This should
// *never* happen while a contenteditable is focused, but as of v28, it
// sometimes does, even when the keypress event target is the contenteditable.
// This breaks the input. Special case these characters to ensure that when
// they are typed, we prevent default on the event to make sure not to
// trigger quickfind.
const FF_QUICKFIND_CHAR = "'";
const FF_QUICKFIND_LINK_CHAR = '/';
const isFirefox = UserAgent.isBrowser('Firefox');

function mustPreventDefaultForCharacter(character: string): boolean {
  return (
    isFirefox &&
    (character == FF_QUICKFIND_CHAR || character == FF_QUICKFIND_LINK_CHAR)
  );
}

/**
 * Replace the current selection with the specified text string, with the
 * inline style and entity key applied to the newly inserted text.
 */
function replaceText(
  editorState: EditorState,
  text: string,
  inlineStyle: DraftInlineStyle,
  entityKey: string | null,
  forceSelection: boolean,
): EditorState {
  const contentState = DraftModifier.replaceText(
    editorState.currentContent,
    editorState.selection,
    text,
    inlineStyle,
    entityKey,
  );
  return pushContent(
    editorState,
    contentState,
    'insert-characters',
    forceSelection,
  );
}

/**
 * When `onBeforeInput` executes, the browser is attempting to insert a
 * character into the editor. Apply this character data to the document,
 * allowing native insertion if possible.
 *
 * Native insertion is encouraged in order to limit re-rendering and to
 * preserve spellcheck highlighting, which disappears or flashes if re-render
 * occurs on the relevant text nodes.
 */
export default function editOnBeforeInput(
  editor: DraftEditor,
  e: SyntheticInputEvent,
): void {
  if (editor._pendingStateFromBeforeInput !== undefined) {
    editor.update(editor._pendingStateFromBeforeInput);
    editor._pendingStateFromBeforeInput = undefined;
  }

  const editorState = editor._latestEditorState;

  const chars: string | null | undefined = (e as any).data;

  // In some cases (ex: IE ideographic space insertion) no character data
  // is provided. There's nothing to do when this happens.
  if (!chars) {
    return;
  }

  // Allow the top-level component to handle the insertion manually. This is
  // useful when triggering interesting behaviors for a character insertion,
  // Simple examples: replacing a raw text ':)' with a smile emoji or image
  // decorator, or setting a block to be a list item after typing '- ' at the
  // start of the block.
  if (
    editor.props.handleBeforeInput &&
    isEventHandled(
      editor.props.handleBeforeInput(chars, editorState, e.timeStamp),
    )
  ) {
    e.preventDefault();
    return;
  }

  // If selection is collapsed, conditionally allow native behavior. This
  // reduces re-renders and preserves spellcheck highlighting. If the selection
  // is not collapsed, we will re-render.
  const selection = editorState.selection;
  const selectionStart = getStartOffset(selection);
  const anchorKey = selection.anchorKey;

  if (!isCollapsed(selection)) {
    e.preventDefault();
    editor.update(
      replaceText(
        editorState,
        chars,
        getCurrentInlineStyle(editorState),
        getEntityKeyForSelection(
          editorState.currentContent,
          editorState.selection,
        ),
        true,
      ),
    );
    return;
  }

  let newEditorState = replaceText(
    editorState,
    chars,
    getCurrentInlineStyle(editorState),
    getEntityKeyForSelection(editorState.currentContent, editorState.selection),
    false,
  );

  // Bunch of different cases follow where we need to prevent native insertion.
  let mustPreventNative = false;
  if (!mustPreventNative) {
    // Browsers tend to insert text in weird places in the DOM when typing at
    // the start of a leaf, so we'll handle it ourselves.
    mustPreventNative = isSelectionAtLeafStart(
      editor._latestCommittedEditorState,
    );
  }
  if (!mustPreventNative) {
    // Let's say we have a decorator that highlights hashtags. In many cases
    // we need to prevent native behavior and rerender ourselves --
    // particularly, any case *except* where the inserted characters end up
    // anywhere except exactly where you put them.
    //
    // Using [] to denote a decorated leaf, some examples:
    //
    // 1. 'hi #' and append 'f'
    // desired rendering: 'hi [#f]'
    // native rendering would be: 'hi #f' (incorrect)
    //
    // 2. 'x [#foo]' and insert '#' before 'f'
    // desired rendering: 'x #[#foo]'
    // native rendering would be: 'x [##foo]' (incorrect)
    //
    // 3. '[#foobar]' and insert ' ' between 'foo' and 'bar'
    // desired rendering: '[#foo] bar'
    // native rendering would be: '[#foo bar]' (incorrect)
    //
    // 4. '[#foo]' and delete '#' [won't use this beforeinput codepath though]
    // desired rendering: 'foo'
    // native rendering would be: '[foo]' (incorrect)
    //
    // 5. '[#foo]' and append 'b'
    // desired rendering: '[#foob]'
    // native rendering would be: '[#foob]'
    // (native insertion here would be ok for decorators like simple spans,
    // but not more complex decorators. To be safe, we need to prevent it.)
    //
    // It is safe to allow native insertion if and only if the full list of
    // decorator ranges matches what we expect native insertion to give, and
    // the range lengths have not changed. We don't need to compare the content
    // because the only possible mutation to consider here is inserting plain
    // text and decorators can't affect text content.
    const oldBlockTree = getBlockTree(editorState, anchorKey);
    const newBlockTree = getBlockTree(newEditorState, anchorKey);
    mustPreventNative =
      oldBlockTree.length !== newBlockTree.length ||
      some(zip(oldBlockTree, newBlockTree), ([oldLeafSet, newLeafSet]) => {
        // selectionStart is guaranteed to be selectionEnd here
        const oldStart = oldLeafSet.start;
        const adjustedStart =
          oldStart + (oldStart >= selectionStart ? chars.length : 0);
        const oldEnd = oldLeafSet.end;
        const adjustedEnd =
          oldEnd + (oldEnd >= selectionStart ? chars.length : 0);
        const newStart = newLeafSet.start;
        const newEnd = newLeafSet.end;
        const newDecoratorKey = newLeafSet.decoratorKey;
        return (
          // Different decorators
          oldLeafSet.decoratorKey !== newDecoratorKey ||
          // Different number of inline styles
          oldLeafSet.leaves.length !== newLeafSet.leaves.length ||
          // Different effective decorator position
          adjustedStart !== newStart ||
          adjustedEnd !== newEnd ||
          // Decorator already existed and its length changed
          (newDecoratorKey != null && newEnd - newStart !== oldEnd - oldStart)
        );
      });
  }
  if (!mustPreventNative) {
    mustPreventNative = mustPreventDefaultForCharacter(chars);
  }
  if (!mustPreventNative) {
    mustPreventNative =
      nullthrows(newEditorState.directionMap).get(anchorKey) !==
      nullthrows(editorState.directionMap).get(anchorKey);
  }

  if (mustPreventNative) {
    e.preventDefault();
    newEditorState = setEditorState(newEditorState, {
      forceSelection: true,
    });
    editor.update(newEditorState);
    return;
  }

  // We made it all the way! Let the browser do its thing and insert the char.
  newEditorState = setEditorState(newEditorState, {
    nativelyRenderedContent: newEditorState.currentContent,
  });
  // The native event is allowed to occur. To allow user onChange handlers to
  // change the inserted text, we wait until the text is actually inserted
  // before we actually update our state. That way when we rerender, the text
  // we see in the DOM will already have been inserted properly.
  editor._pendingStateFromBeforeInput = newEditorState;
  setImmediate(() => {
    if (editor._pendingStateFromBeforeInput !== undefined) {
      editor.update(editor._pendingStateFromBeforeInput);
      editor._pendingStateFromBeforeInput = undefined;
    }
  });
}
