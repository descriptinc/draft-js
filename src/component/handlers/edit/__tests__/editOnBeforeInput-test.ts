/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+draft_js
 * @format
 */

import {makeSelectionState} from '../../../../model/immutable/SelectionState';
import {
  acceptSelection,
  createWithContent,
  setEditorState,
} from '../../../../model/immutable/EditorState';
import {
  ContentState,
  createFromBlockArray,
} from '../../../../model/immutable/ContentState';
import {
  ContentBlock,
  makeContentBlock,
} from '../../../../model/immutable/ContentBlock';
import DraftEditor from '../../../base/DraftEditor.react';
import {SyntheticInputEvent} from '../../../utils/eventTypes';
import editOnBeforeInput from '../editOnBeforeInput';
import CompositeDraftDecorator from '../../../../model/decorators/CompositeDraftDecorator';

const DEFAULT_SELECTION = {
  anchorKey: 'a',
  anchorOffset: 0,
  focusKey: 'a',
  focusOffset: 0,
  isBackward: false,
};

const rangedSelection = makeSelectionState({
  ...DEFAULT_SELECTION,
  focusOffset: 1,
});

const rangedSelectionBackwards = makeSelectionState({
  ...DEFAULT_SELECTION,
  anchorOffset: 1,
  isBackward: true,
});

const getEditorState = (text: string = 'Arsenal') => {
  return createWithContent(
    createFromBlockArray([
      makeContentBlock({
        key: 'a',
        text,
      }),
    ]),
  );
};

const getDraftEditor = (obj): DraftEditor => obj as any;

const getInputEvent = (data?: any): SyntheticInputEvent =>
  ({
    data,
    preventDefault: jest.fn(),
  } as any);

test('editor is not updated if no character data is provided', () => {
  const editorState = acceptSelection(getEditorState(), rangedSelection);

  const editor = getDraftEditor({
    _latestEditorState: editorState,
    props: {},
    update: jest.fn(),
  });

  editOnBeforeInput(editor, getInputEvent());

  expect(editor.update).toHaveBeenCalledTimes(0);
});

test('editor is not updated if handled by handleBeforeInput', () => {
  const editorState = acceptSelection(getEditorState(), rangedSelection);

  const editor = getDraftEditor({
    _latestEditorState: editorState,
    props: {
      handleBeforeInput: () => true,
    },
    update: jest.fn(),
  });

  editOnBeforeInput(editor, getInputEvent('O'));

  expect(editor.update).toHaveBeenCalledTimes(0);
});

test('editor is updated with new text if it does not match current selection', () => {
  const editorState = acceptSelection(getEditorState(), rangedSelection);

  const update = jest.fn();
  const editor = getDraftEditor({
    _latestEditorState: editorState,
    props: {},
    update,
  });

  editOnBeforeInput(editor, getInputEvent('O'));

  expect(update).toHaveBeenCalledTimes(1);

  const newEditorState = update.mock.calls[0][0];
  expect(newEditorState.currentContent).toMatchSnapshot();
});

test('editor selectionstate is updated if new text matches current selection', () => {
  const editorState = acceptSelection(getEditorState(), rangedSelection);

  const update = jest.fn();
  const editor = getDraftEditor({
    _latestEditorState: editorState,
    props: {},
    update,
  });

  editOnBeforeInput(editor, getInputEvent('A'));

  expect(update).toHaveBeenCalledTimes(1);

  const newEditorState = update.mock.calls[0][0];
  expect(newEditorState.selection).toMatchSnapshot();
});

test('editor selectionstate is updated if new text matches current selection and user selected backwards', () => {
  const editorState = acceptSelection(
    getEditorState(),
    rangedSelectionBackwards,
  );

  const update = jest.fn();
  const editor = getDraftEditor({
    _latestEditorState: editorState,
    props: {},
    update,
  });

  editOnBeforeInput(editor, getInputEvent('A'));

  expect(update).toHaveBeenCalledTimes(1);

  const newEditorState = update.mock.calls[0][0];
  expect(newEditorState.selection).toMatchSnapshot();
});

const HASHTAG_REGEX = /#[a-z]+/g;
function hashtagStrategy(
  contentBlock: ContentBlock,
  callback,
  _contentState: ContentState,
) {
  findWithRegex(HASHTAG_REGEX, contentBlock, callback);
}

function findWithRegex(regex: RegExp, contentBlock: ContentBlock, callback) {
  const text = contentBlock.text;
  let matchArr = regex.exec(text);
  while (matchArr !== null) {
    callback(matchArr.index, matchArr.index + matchArr[0].length);
    matchArr = regex.exec(text);
  }
}

function testDecoratorFingerprint(
  text: string,
  selection: number,
  charToInsert: string,
  shouldPrevent,
) {
  const editorState = acceptSelection(
    setEditorState(getEditorState(text), {
      decorator: new CompositeDraftDecorator([
        {
          strategy: hashtagStrategy,
          component: () => null,
        },
      ]),
    }),
    makeSelectionState({
      ...DEFAULT_SELECTION,
      anchorOffset: selection,
      focusOffset: selection,
    }),
  );

  const editor = getDraftEditor({
    _latestEditorState: editorState,
    _latestCommittedEditorState: editorState,
    props: {},
    update: jest.fn(),
  });

  const ev = getInputEvent(charToInsert);
  editOnBeforeInput(editor, ev);

  expect(ev.preventDefault.mock.calls.length).toBe(shouldPrevent ? 1 : 0);
}

test('decorator fingerprint logic bails out of native insertion', () => {
  const oldGetSelection = global.getSelection;
  try {
    global.getSelection = () => ({});

    // Make sure we prevent native insertion in the right cases
    testDecoratorFingerprint('hi #', 4, 'f', true);
    testDecoratorFingerprint('x #foo', 3, '#', true);
    testDecoratorFingerprint('#foobar', 4, ' ', true);
    testDecoratorFingerprint('#foo', 4, 'b', true);
    testDecoratorFingerprint('#foo bar #baz', 2, 'o', true);
    testDecoratorFingerprint('#foo bar #baz', 12, 'a', true);

    // but these are OK to let through
    testDecoratorFingerprint('#foo bar #baz', 7, 'o', false);
    testDecoratorFingerprint('start #foo bar #baz end', 5, 'a', false);
    testDecoratorFingerprint('start #foo bar #baz end', 20, 'a', false);
  } finally {
    global.getSelection = oldGetSelection;
  }
});
