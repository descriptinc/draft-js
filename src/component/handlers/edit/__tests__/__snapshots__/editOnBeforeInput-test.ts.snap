// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`editor is updated with new text if it does not match current selection 1`] = `
Object {
  "blockMap": Map {
    "a" => Object {
      "characterList": Array [
        Object {
          "entity": null,
          "style": Set {},
        },
        Object {
          "entity": null,
          "style": Set {},
        },
        Object {
          "entity": null,
          "style": Set {},
        },
        Object {
          "entity": null,
          "style": Set {},
        },
        Object {
          "entity": null,
          "style": Set {},
        },
        Object {
          "entity": null,
          "style": Set {},
        },
        Object {
          "entity": null,
          "style": Set {},
        },
      ],
      "data": Object {},
      "depth": 0,
      "key": "a",
      "text": "Orsenal",
      "type": "unstyled",
    },
  },
  "selectionAfter": Object {
    "anchorKey": "a",
    "anchorOffset": 1,
    "focusKey": "a",
    "focusOffset": 1,
    "hasFocus": false,
    "isBackward": false,
  },
  "selectionBefore": Object {
    "anchorKey": "a",
    "anchorOffset": 0,
    "focusKey": "a",
    "focusOffset": 1,
    "hasFocus": false,
    "isBackward": false,
  },
}
`;

exports[`editor selectionstate is updated if new text matches current selection 1`] = `
Object {
  "anchorKey": "a",
  "anchorOffset": 1,
  "focusKey": "a",
  "focusOffset": 1,
  "hasFocus": false,
  "isBackward": false,
}
`;

exports[`editor selectionstate is updated if new text matches current selection and user selected backwards 1`] = `
Object {
  "anchorKey": "a",
  "anchorOffset": 1,
  "focusKey": "a",
  "focusOffset": 1,
  "hasFocus": false,
  "isBackward": false,
}
`;
